ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Block.hs

```diff
@@ -0,0 +1,243 @@
+fromShelleyPrevHash :: SL.PrevHash c -> ChainHash (Header (ShelleyBlock c))
+fromShelleyPrevHash SL.GenesisHash   = GenesisHash
+fromShelleyPrevHash (SL.BlockHash h) = BlockHash (ShelleyHash h)
+
+-- | From @ouroboros-consensus@ to @cardano-ledger-specs@
+toShelleyPrevHash :: ChainHash (Header (ShelleyBlock c)) -> SL.PrevHash c
+toShelleyPrevHash GenesisHash                 = SL.GenesisHash
+toShelleyPrevHash (BlockHash (ShelleyHash h)) = SL.BlockHash h
+
+{-------------------------------------------------------------------------------
+  Serialisation
+-------------------------------------------------------------------------------}
+
+instance Crypto c => ToCBOR (ShelleyBlock c) where
+  -- Don't encode the header hash, we recompute it during deserialisation
+  toCBOR = toCBOR . shelleyBlockRaw
+
+instance Crypto c => FromCBOR (ShelleyBlock c) where
+  fromCBOR = mkShelleyBlock <$> fromCBOR
+
+instance Crypto c => ToCBOR (Header (ShelleyBlock c)) where
+  -- Don't encode the header hash, we recompute it during deserialisation
+  toCBOR = toCBOR . shelleyHeaderRaw
+
+instance Crypto c => FromCBOR (Header (ShelleyBlock c)) where
+  fromCBOR = mkShelleyHeader <$> fromCBOR
+
+encodeShelleyBlockWithInfo :: Crypto c => ShelleyBlock c -> BinaryInfo Encoding
+encodeShelleyBlockWithInfo blk = BinaryInfo {
+      binaryBlob   = toCBOR blk
+      -- See 'shelleyAddHeaderEnvelope'
+    , headerOffset = 0
+      -- TODO When we have annotations in the Shelley decoders, this will
+      -- become much cheaper
+    , headerSize   = fromIntegral $ Lazy.length (serialize (getHeader blk))
+    }
+
+-- | When given the raw header bytes extracted from the block (see
+-- 'encodeShelleyBlockWithInfo'), we still need to prepend the correct list
+-- length so that we can decode the header with 'fromCBOR':
+--
+-- > > putStrLn $ prettyHexEnc $ toCBOR blk
+-- > 93  # list(19)
+-- >    44 ec 7f 7e 7b  # bytes(4)
+-- >    1a 20 28 e0 2b  # int(539549739)
+-- >    ...
+-- >         0a  # integer(10)
+-- >    80  # list(0)
+-- >    80  # list(0)
+-- >    a0  # map(0)
+--
+-- > > putStrLn $ prettyHexEnc $ toCBOR (getHeader blk)
+-- > 90  # list(16)
+-- >    44 ec 7f 7e 7b  # bytes(4)
+-- >    1a 20 28 e0 2b  # int(539549739)
+-- >    ...
+-- >          0a  # integer(10)
+--
+-- Note that the block contains three fields more than the header. For some
+-- reason, the header is not nested in the block, but its \"list\"
+-- ('encodeListLen') is extended with the three extra fields.
+--
+-- This function will be given the bytestring corresponding to the encoding of
+-- the block with but without the three extra fields that are not part of the
+-- header. This function will drop the first byte (\"list length of 19\") and
+-- prepend a byte representing a \"list length of 16\" and leave the remainder
+-- of the bytestring untouched.
+shelleyAddHeaderEnvelope :: Lazy.ByteString -> Lazy.ByteString
+shelleyAddHeaderEnvelope = Lazy.cons 0x90 . Lazy.drop 1
```

Comment:
This will change with https://github.com/input-output-hk/cardano-ledger-specs/pull/1313/